src/gui/qgsrelationeditorwidget.cpp

```diff
@@ -487,9 +488,84 @@ void QgsRelationEditorWidget::deleteSelectedFeatures()
 
 void QgsRelationEditorWidget::deleteFeatures( const QgsFeatureIds &featureids )
 {
-  QgsVectorLayer *layer = mNmRelation.isValid() ? mNmRelation.referencedLayer() : mRelation.referencingLayer();
-  layer->deleteFeatures( featureids );
-  updateUi();
+  bool deleteFeatures = true;
+
+  QgsVectorLayer *layer;
+  if ( mNmRelation.isValid() )
+  {
+    layer = mNmRelation.referencedLayer();
+
+    // When deleting a linked feature within an N:M relation,
+    // check if the feature is linked to more than just one feature.
+    // In case it is linked more than just once, ask the user for confirmation
+    // as it is likely he was not aware of the implications and might either
+    // leave the dataset in a corrupted state (referential integrity) or if
+    // the fk constraint is ON CASCADE DELETE, there may be several linking
+    // entries deleted along.
+
+    QgsFeatureRequest deletedFeaturesRequest;
+    deletedFeaturesRequest.setFilterFids( featureids );
+    deletedFeaturesRequest.setFlags( QgsFeatureRequest::NoGeometry );
+    deletedFeaturesRequest.setSubsetOfAttributes( QgsAttributeList() << mNmRelation.referencedFields().first() );
+
+    QgsFeatureIterator deletedFeatures = layer->getFeatures( deletedFeaturesRequest );
```

Comment:
Could we check that `layer` is not a `nullptr` before using it? You know, just to be on the safe side :).

Reply:
I thought about it, but then again, `QgsRelation::isValid()` already checks those

Reply:
Generally speaking, I prefer an explicit code. Moreover, if someday `QgsRelation::isValid()` is badly modified, we'll have a possible source of error. 

"Anyway, perfection is boring!"  ;)

Sorry about the noise :).

Reply:
Good thinking, but if we want to be explicit and on the safe side we even need to use guarded pointers. Do you think it's worth the effort or should we trust the environment (attribute form I guess) to clean things up when a layer is deleted?

Reply:
I believe that @nyalldawson often uses `QPointer`, but I don't know if it's somehting that we want to generalise?

Reply:
This is mostly done for long living objects which are used in a wide variety of situations (classic example, pointers in expression contexts). And actually, I think if we talk at `QgsRelation` level (and not widget level) it could very well be treated as something like this.